### PR TITLE
HDDS-5455. Check invalid keyName when creating new key

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

In OzoneManagerStateMachine.runCommand, if there is a Runtime exception, OM will be terminated.

In our case, the user would like to create an object with malformed characters in keyName, and the following error happened and throws an InvalidPathException, which is a Runtime Exception causing the OM's shutdown.

Normally this wouldn't happen, since a user can not easily create an invalid keyName with UTF-8. It happened in our environment because our OM started with the property of "file.encoding" and "sun.jnu.encoding" set to "ANSI_X3.4-1968", and it's easy for the user to write a invalid keyName.

This issue can be avoided when running in a good environment, but it's still not reasonable that a client create key operation causing the shutdown of OM.

This ticket is to check the keyName if it fits the current encoding of OM, if not just return the OMException of "INVALID_KEY_NAME".

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5455

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by symious** [comment]
@elek @adoroszlai Could you help to check this PR?

**Review by adoroszlai — COMMENTED** [review]
Thanks @symious for reporting this problem.

I have some concerns about the proposed fix:

 1. I feel that `Paths.get` is an implementation detail.  Most key names that trigger the exception are not inherently invalid.  We should try to avoid restricting key names just to make `Paths.get` happy.  Does restarting OM with a different default encoding possibly make existing keys invalid?  OM already has a check for valid key names, but it's optional: https://github.com/apache/ozone/blob/d4043397d566c4a1b49ba06264b2b94b127562ff/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java#L89-L95
 2. Other requests might also be affected by this problem.  At least it can be reproduced in `OMDirectoryCreateRequest` with `ozone fs -mkdir`.  Possibly others, too.

**Comment by symious** [comment]
@adoroszlai Thanks for the view.

Seems the config of "OZONE_OM_KEYNAME_CHARACTER_CHECK_ENABLED_KEY" will deny
```
   * "\", "{", "}", "^", "<", ">", "#", "|", "%", "`", "[", "]", "~", "?"
   * and Non-printable ASCII characters (128–255 decimal characters).
```
, but if the user does create a name contains something like "U-00000000", it won't help with the issue.

I think it is the catastrophic outcome that is not acceptable to users. In our scenario, SRE provides the basic env and we install the applications, we are not quite sure if SRE has set the correcting encoding. So we start the Ozone cluster, then it runs well for several months, suddenly the OM crashes because users are creating some invalid keys. 

Besides, the keys are created by the scheduling system, the invalid key was created several hours after we told them to correct them, which means if we don't check the "Paths.get", the OM cluster isn't available for several hours, and it's not even decided by cluster maintainers, but the users.

It took some time before I found out it's related to the encoding. When the outage happens, and the error message is so obvious, I won't even check the env, so the recovery still relies on the users.

IMHO, the encoding of JVM should affect Ozone so much. If Ozone runs on "ANSI_X3.4-1968", it should provide the functions on this encoding, and if this encoding can not resolve a keyName, it should just return an INVALID_KEY_NAME and reject. If runs on "UTF-8", it should reject the user requests as well if it can not resolve. 


Attached the original error log and keyName.
The original error is as follows:
```
java.nio.file.InvalidPathException: Malformed input or input contains unmappable characters: /vol_name/bucket_name/task-exec/di_scheduler_test.monhly_dudutestshell063002??prod_2021071519_HOUR_1_0_prod/main.sh
	at sun.nio.fs.UnixPath.encode(UnixPath.java:147)
	at sun.nio.fs.UnixPath.<init>(UnixPath.java:71)
	at sun.nio.fs.UnixFileSystem.getPath(UnixFileSystem.java:281)
	at java.nio.file.Paths.get(Paths.java:84)
	at org.apache.hadoop.ozone.util.RadixTree.getLongestPrefix(RadixTree.java:190)
	at org.apache.hadoop.ozone.om.PrefixManagerImpl.getLongestPrefixPath(PrefixManagerImpl.java:260)
	at org.apache.hadoop.ozone.om.request.key.OMKeyRequest.getAclsForKey(OMKeyRequest.java:284)
	at org.apache.hadoop.ozone.om.request.key.OMKeyRequest.createKeyInfo(OMKeyRequest.java:267)
	at org.apache.hadoop.ozone.om.request.key.OMKeyRequest.prepareKeyInfo(OMKeyRequest.java:352)
	at org.apache.hadoop.ozone.om.request.key.OMKeyCreateRequest.validateAndUpdateCache(OMKeyCreateRequest.java:278)
	at org.apache.hadoop.ozone.protocolPB.OzoneManagerRequestHandler.handleWriteRequest(OzoneManagerRequestHandler.java:227)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerStateMachine.runCommand(OzoneManagerStateMachine.java:417)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerStateMachine.lambda$applyTransaction$1(OzoneManagerStateMachine.java:242)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
The keyName user wants to create is `task-exec/di_scheduler_test.monhly_dudutestshell063002\342\200\224\342\200\224prod_2021071519_HOUR_1_0_prod/main.sh`.

**Comment by adoroszlai** [comment]
> I think it is the catastrophic outcome that is not acceptable to users.

I agree and would like to get this fixed.  I'm just wondering about the best way.  I would prefer least intrusion for users, even if it requires more work now code-wise.  For example we might want to replace existing `Paths.get` usage with some other implementation insensitive to `file.encoding`.  If I understand correctly it is mainly used as a convenient way to split key names at `/`, not for actually creating filesystem objects.

**Comment by symious** [comment]
Seems `Paths.get()` has been used in many places other than `RadixTree.getLongestPrefix()`.

**Comment by adoroszlai** [comment]
> Seems `Paths.get()` has been used in many places other than `RadixTree.getLongestPrefix()`.

I guess we should fix those places, too.

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Revisiting this PR:

1. I still think using `Paths.get()` for string splitting is an implementation detail that should not affect users.
2. What if OM is restarted with a different encoding?  Previously valid key name suddenly becomes invalid?
3. Several other OM requests are also affected (file create, dir create, same `...WithFSO`, S3 multipart).

**Comment by symious** [comment]
@adoroszlai Thanks for revisiting this PR.
I was wondering that this ticket is more than file encoding. As the test case shows, it can be triggered deliberately, a malicious user could be able to crash the system with invalid keyname.

**Comment by adoroszlai** [comment]
> @adoroszlai it can be triggered deliberately, a malicious user could be able to crash the system with invalid keyname.

Understood.  That's why we should fix it.  Just not by rejecting requests.

**Comment by symious** [comment]
If there is a invalid character, it will always throw InvalidPathException if we use methods related to "Path".
So are you suggesting that we should remove all the use of "Path", and simply use String to do the work?
If so this could be a big change, since many interfaces are using "Path", many compatible issues might happen during the change.

**Comment by adoroszlai** [comment]
> So are you suggesting that we should remove all the use of "Path", and simply use String to do the work?

Exactly.  Only for volume/bucket/key name parsing, we can keep Path for real file-related operations.

**Comment by symious** [comment]
Sorry for the late reply. I tried to replace all the "Paths.get" with "String.split" and array oprerations, but I found some special cases  with "~", ".", "..", which needs the help from "Paths", I'm not sure if it's the best approach to do the replacement here.

**Comment by adoroszlai** [comment]
> I found some special cases with "~", ".", "..", which needs the help from "Paths"

Why do we need any of those for volume/bucket/key names?

**Comment by symious** [comment]
I think it's normal when users are using OFS protocol.

**Comment by kerneltime** [comment]
@symious does this PR need to be revisited?

**Comment by symious** [comment]
@kerneltime Sure, thank you for the review.

I updated the PR, because in our cluster, we found invalidPathException not only in CreateKeyRequest, it also emerges in other requests like CreateFileRequset, DeleteKeyRequset.

The trick thing is the path from Protobuf message seems normal, but sometimes it still throws the exception, and it can not be reproduced, it seems happen only once every two days in our UAT cluster.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:317]: I don't think we should log to OM log client-induced errors. This could be something logged to the audit log instead for a failed request.

Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java]: @kerneltime Thank you for the review. Updated the PR, please have a look.

**Comment by adoroszlai** [comment]
@symious Please try to avoid force-push when updating the PR.  Here are some great articles that explain why:

https://developers.mattermost.com/blog/submitting-great-prs/#4-avoid-force-pushing
https://www.freecodecamp.org/news/optimize-pull-requests-for-reviewer-happiness#request-a-review

**Comment by symious** [comment]
@adoroszlai Understand, thank you for the reminder.

**Comment by swamirishi** [comment]
@symious @kerneltime @adoroszlai can this be merged?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:330]: I am not sure why we need to do this

Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:330]: There was a inconsistence of the transactionId among the OMs, followers have problem starting without these lines.

Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:393]: @symious , thanks for continue working on this patch.  Can you explain it a bit more about why we need to add this omClientResponse.getFlushFuture() check?

Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:392]: I think I added it because the flustFuture might be already set in "validateAndUpdateCache()", so if it's already set, we don't need to do it again.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/InvalidPathClientResponse.java
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/InvalidPathClientResponse.java:32]: Can we specify a more specific table list to cleanup ?

**Comment by ChenSammi** [comment]
@symious , could you add a test case for this fix?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:392]: When InvalidPathException is handled in handleWriteRequest, instead of validateAndUpdateCache, there will be no audit log at all for this request. Since audit log is important, I would suggest move the InvalidPathException into validateAndUpdateCache if possible.

**Comment by symious** [comment]
@ChenSammi Added exception to each request case, and test case added, PTAL.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/InvalidPathClientResponse.java
Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/InvalidPathClientResponse.java]: Since `InvalidPathClientResponse` not being used now, removed the class.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketCreateRequest.java
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketCreateRequest.java:254]: Can this exception translation logic be moved into `createErrorOMResponse`?  It would help avoid code duplication, and I think it's also better if the code using `exception` (logging, etc.) gets the original one.

**Review by adoroszlai — COMMENTED** [review]
Thanks a lot @symious for continued efforts on this problem.

**Comment by ChenSammi** [comment]
@symious , can you check the checkstyle report? there are unused imports "import java.nio.file.InvalidPathException;" .

**Comment by adoroszlai** [comment]
Shouldn't we catch `InvalidPathException` and report `INVALID_PATH` in the response?

To clarify my previous suggestion:

```
    } catch (IOException | InvalidPathException ex) {
      exception = ex instanceof IOException ? (IOException) ex :
          new OMException(ex.getMessage(),
              OMException.ResultCodes.INVALID_PATH);
      omClientResponse = new OMBucketCreateResponse(
          createErrorOMResponse(omResponse, exception));
```

should be:

```
    } catch (IOException | InvalidPathException ex) {
      exception = ex;
      omClientResponse = new OMBucketCreateResponse(
          createErrorOMResponse(omResponse, ex));
```

`createErrorOMResponse` (and `exceptionToResponseStatus`) should handle `InvalidPathException`, setting `INVALID_PATH`  as response status.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketAclRequest.java
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketAclRequest.java:139]: Can we revert this change? Also many of below request java files, which change the "ex" to "exception".

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java:355]: Also please revert this change.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java:150]: Can we check here if the exception cause is InvalidPathException, 
and then turn it to an OMException with INVALID_PATH error code.

```
 new OMException(ex.getMessage(),
              OMException.ResultCodes.INVALID_PATH);
```

Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java]: Updated.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java:353]: @symious , this one can be reverted too.

Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java]: Sure, updated.

**Comment by ChenSammi** [comment]
Let's wait for the CI to pass.  The last patch LGTM.

**Comment by ChenSammi** [comment]
The failed TestOMKeyCreateRequestWithFSO seems related. @symious , could you have a look ?

**Comment by symious** [comment]
> The failed TestOMKeyCreateRequestWithFSO seems related. @symious , could you have a look ?

@ChenSammi Currently we only catch the InvalidPathException from RadixTree, but in some FSO classes, "Paths.get" is direclty used, the error related to the failed case is from "https://github.com/apache/ozone/blob/33140cc3dab284c7f830770c22b438d0dbb3a076/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequestWithFSO.java#L123".

If we catch the exception for FSO classes, it will fall back to the earlier solution.

**Comment by ChenSammi** [comment]
> > The failed TestOMKeyCreateRequestWithFSO seems related. @symious , could you have a look ?
> 
> @ChenSammi Currently we only catch the InvalidPathException from RadixTree, but in some FSO classes, "Paths.get" is direclty used, the error related to the failed case is from
> 
> https://github.com/apache/ozone/blob/33140cc3dab284c7f830770c22b438d0dbb3a076/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequestWithFSO.java#L123
> 
> If we catch the exception for FSO classes, it will fall back to the earlier solution.

How about we provide a Ozone version Paths.get(), and replace the existing code with this one?

**Comment by symious** [comment]
> @symious , I see. How about we provide a Ozone version Paths.get(), and replace the existing code with this one?

I'm not sure if it's a good idea, the original one seems a common used one, the only flaw in Ozone I think is that OM didn't handle the exception properly and terminates the process directly. Maybe we can try get back to the earlier solution and try @adoroszlai 's advice?

**Comment by symious** [comment]
@ChenSammi @adoroszlai    Updatd the PR, PTAL.

**Review by adoroszlai — COMMENTED** [review]
Thanks @symious for updating the patch, LGTM.

**Review by ChenSammi — APPROVED** [review]
Thanks @symious for the contribution, and @adoroszlai for review.

**Pull request merged**
